File: packages/engine/src/onboarding.ts
/**
 * Onboarding sensoriel-first — state machine + identity guard.
 *
 * CDC ref : F-012 (Onboarding sensoriel AVANT identité, 3 écrans theme → motion → density).
 * Brick   : B-013.
 * Risk    : Critical 95% + MC/DC + PBT (Dignity §a L'ACCUEIL BLOCKING).
 *
 * Spec :
 *   - State machine: idle → started → (theme → motion → density) → completed.
 *   - `canCollectIdentity()` returns true ONLY when all 3 sensoriel steps done.
 *     This is the Dignity §a BLOCKING contract: ZÉRO identité avant sensoriel.
 *   - `completeStep(step, value)` writes the user value to the root prefs key
 *     under MORPHIC_STORAGE_KEY (consumed by morphicInit / setTheme / etc.).
 *   - `skipStep(step)` applies a documented default (theme='auto', motion='auto',
 *     density='comfortable') so a user can pass through without choosing.
 *   - Order is enforced: completeStep/skipStep on a step that is not the
 *     currentStep throws.
 *   - Events: `morphic:onboarding:step-complete` with `{step, value}` after each
 *     step ; `morphic:onboarding:complete` after the final step (density).
 *
 * Defensive contracts (≥2 per critical function per PET §5) :
 *   - Invalid step name → TypeError (poka-yoke, before any state read).
 *   - Non-string value for completeStep → TypeError.
 *   - completeStep/skipStep before startOnboarding → Error "not started".
 *   - Out-of-order or post-completion call → Error "invalid order".
 *   - State NOT mutated when validation fails (snapshot-stable).
 *   - localStorage failures (read or write) do NOT throw — in-memory state wins.
 *   - SSR-safe: document / CustomEvent / localStorage guards.
 *   - Storage corruption-tolerant: malformed JSON, non-object, or out-of-shape
 *     sub-key all fall back to idle.
 *
 * Persistence layout : the onboarding state lives under sub-key
 * `MORPHIC_ONBOARDING_MARKER` inside `MORPHIC_STORAGE_KEY`. The user values
 * (theme/motion/density) are written to the root keys so morphicInit picks
 * them up on next page load. `resetOnboarding()` only clears the sub-key —
 * the user-written prefs are preserved (RGPD-friendly: erase ceremony, not
 * choices).
 */

import { MORPHIC_STORAGE_KEY } from './init.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Ordered list of sensoriel onboarding steps. */
export const ONBOARDING_STEPS = ['theme', 'motion', 'density'] as const;

/** Step name (closed enum). */
export type OnboardingStep = (typeof ONBOARDING_STEPS)[number];

/** Storage sub-key under MORPHIC_STORAGE_KEY for onboarding state. */
export const MORPHIC_ONBOARDING_MARKER = 'morphic-onboarding' as const;

/** Event dispatched after each step completion (complete or skip). */
export const MORPHIC_ONBOARDING_EVENT_STEP_COMPLETE = 'morphic:onboarding:step-complete' as const;

/** Event dispatched once all 3 steps are done. */
export const MORPHIC_ONBOARDING_EVENT_COMPLETE = 'morphic:onboarding:complete' as const;

/** Default value applied by `skipStep` for each axis. */
const STEP_DEFAULTS: Record<OnboardingStep, string> = {
  theme: 'auto',
  motion: 'auto',
  density: 'comfortable',
};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Runtime state of the onboarding flow. */
export interface OnboardingState {
  /** True once the user has entered the flow at least once. */
  started: boolean;
  /** True once all 3 steps have been completed (complete or skip). */
  completed: boolean;
  /** Current step waiting for user input, or null when idle/done. */
  currentStep: OnboardingStep | null;
  /** Steps completed so far, in order. */
  completedSteps: OnboardingStep[];
}

// ---------------------------------------------------------------------------
// In-memory cache
// ---------------------------------------------------------------------------

let currentState: OnboardingState | null = null;

/**
 * Test-only reset hook. Module-level state survives across vitest test cases
 * within the same file (modules are shared per worker). Tests call this in
 * `beforeEach` to guarantee a clean slate.
 */
export function __resetOnboardingStateForTests(): void {
  currentState = null;
}

// ---------------------------------------------------------------------------
// Validation (closed enum — poka-yoke)
// ---------------------------------------------------------------------------

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function isValidStep(value: unknown): value is OnboardingStep {
  return typeof value === 'string' && (ONBOARDING_STEPS as readonly string[]).includes(value);
}

// ---------------------------------------------------------------------------
// Storage I/O (SSR-safe, corruption-tolerant)
// ---------------------------------------------------------------------------

function readRoot(): Record<string, unknown> {
  if (typeof localStorage === 'undefined') return {};
  let raw: string | null;
  try {
    raw = localStorage.getItem(MORPHIC_STORAGE_KEY);
  } catch {
    return {};
  }
  if (raw === null) return {};
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    return {};
  }
  return isPlainObject(parsed) ? parsed : {};
}

function writeRoot(root: Record<string, unknown>): void {
  if (typeof localStorage === 'undefined') return;
  try {
    localStorage.setItem(MORPHIC_STORAGE_KEY, JSON.stringify(root));
  } catch {
    // localStorage unavailable (private mode, quota) — in-memory wins.
  }
}

function deriveCurrentStep(
  completedSteps: readonly OnboardingStep[],
  completed: boolean,
): OnboardingStep | null {
  if (completed) return null;
  const next = ONBOARDING_STEPS[completedSteps.length];
  return next ?? null;
}

function readStoredState(): OnboardingState | null {
  const root = readRoot();
  const stored = root[MORPHIC_ONBOARDING_MARKER];
  if (!isPlainObject(stored)) return null;

  const started = stored.started;
  const completed = stored.completed;
  const rawSteps = stored.completedSteps;
  if (typeof started !== 'boolean') return null;
  if (typeof completed !== 'boolean') return null;
  if (!Array.isArray(rawSteps)) return null;

  const completedSteps = rawSteps.filter((s): s is OnboardingStep => isValidStep(s));
  return {
    started,
    completed,
    completedSteps,
    currentStep: deriveCurrentStep(completedSteps, completed),
  };
}

function getStateInternal(): OnboardingState {
  if (currentState !== null) return currentState;
  const fromStorage = readStoredState();
  if (fromStorage !== null) {
    currentState = fromStorage;
    return fromStorage;
  }
  return { started: false, completed: false, currentStep: null, completedSteps: [] };
}

function persistState(state: OnboardingState, prefWrites?: Record<string, string>): void {
  const root = readRoot();
  root[MORPHIC_ONBOARDING_MARKER] = {
    started: state.started,
    completed: state.completed,
    completedSteps: state.completedSteps,
  };
  if (prefWrites !== undefined) {
    for (const [k, v] of Object.entries(prefWrites)) {
      root[k] = v;
    }
  }
  writeRoot(root);
}

// ---------------------------------------------------------------------------
// Event dispatch (SSR-safe)
// ---------------------------------------------------------------------------

function emit(name: string, detail: Record<string, unknown>): void {
  if (typeof document === 'undefined' || typeof CustomEvent === 'undefined') return;
  document.dispatchEvent(new CustomEvent(name, { detail, bubbles: true }));
}

// ---------------------------------------------------------------------------
// Read API
// ---------------------------------------------------------------------------

/** Return the current onboarding state (in-memory first, then storage). */
export function getOnboardingState(): OnboardingState {
  return getStateInternal();
}

/**
 * Dignity §a BLOCKING contract — return `true` if AND ONLY IF the user has
 * completed all 3 sensoriel steps. Host apps MUST gate any identity-collection
 * UI behind this guard.
 *
 * MC/DC truth table (started AND completed):
 *  - F & F → false (initial idle)
 *  - T & F → false (in-progress)
 *  - T & T → true  (done)
 *  - F & T → impossible (state machine prevents)
 */
export function canCollectIdentity(): boolean {
  const state = getStateInternal();
  return state.started && state.completed;
}

// ---------------------------------------------------------------------------
// Mutation API
// ---------------------------------------------------------------------------

/**
 * Start (or re-enter) the onboarding flow.
 *
 * Idempotent : if already started, returns the existing state unchanged.
 * Does NOT reset completedSteps or completed flag.
 */
export function startOnboarding(): OnboardingState {
  const existing = getStateInternal();
  if (existing.started) return existing;

  const newState: OnboardingState = {
    started: true,
    completed: false,
    currentStep: 'theme',
    completedSteps: [],
  };
  currentState = newState;
  persistState(newState);
  return newState;
}

/**
 * Internal advance — shared logic between `completeStep` and `skipStep`.
 * Caller has already validated `step` is in the enum. State checks below
 * happen BEFORE any mutation (poka-yoke).
 */
function advanceStep(step: OnboardingStep, value: string): OnboardingState {
  const existing = getStateInternal();

  if (!existing.started) {
    throw new Error('onboarding: not started — call startOnboarding() first');
  }
  if (existing.completed) {
    throw new Error(
      `onboarding: invalid order — onboarding already completed, cannot advance "${step}"`,
    );
  }
  if (existing.currentStep !== step) {
    throw new Error(
      `onboarding: invalid order — expected "${String(existing.currentStep)}", got "${step}"`,
    );
  }

  const completedSteps: OnboardingStep[] = [...existing.completedSteps, step];
  const allDone = completedSteps.length === ONBOARDING_STEPS.length;
  const newState: OnboardingState = {
    started: true,
    completed: allDone,
    currentStep: deriveCurrentStep(completedSteps, allDone),
    completedSteps,
  };
  currentState = newState;
  persistState(newState, { [step]: value });

  emit(MORPHIC_ONBOARDING_EVENT_STEP_COMPLETE, { step, value });
  if (allDone) {
    emit(MORPHIC_ONBOARDING_EVENT_COMPLETE, { state: newState });
  }
  return newState;
}

/**
 * Complete the current step with a user-provided value.
 *
 * @throws {TypeError} when `step` is not in the closed enum.
 * @throws {TypeError} when `value` is not a string.
 * @throws {Error} when onboarding is not started, already done, or step is out of order.
 */
export function completeStep(step: OnboardingStep, value: string): OnboardingState {
  if (!isValidStep(step)) {
    throw new TypeError(
      `completeStep: invalid step. Expected one of ${ONBOARDING_STEPS.join(', ')}, got ${String(step)}.`,
    );
  }
  if (typeof (value as unknown) !== 'string') {
    throw new TypeError(`completeStep: value must be a string, got ${typeof (value as unknown)}.`);
  }
  return advanceStep(step, value);
}

/**
 * Skip the current step using its documented default
 * (theme='auto', motion='auto', density='comfortable').
 *
 * @throws {TypeError} when `step` is not in the closed enum.
 * @throws {Error} when onboarding is not started, already done, or step is out of order.
 */
export function skipStep(step: OnboardingStep): OnboardingState {
  if (!isValidStep(step)) {
    throw new TypeError(
      `skipStep: invalid step. Expected one of ${ONBOARDING_STEPS.join(', ')}, got ${String(step)}.`,
    );
  }
  return advanceStep(step, STEP_DEFAULTS[step]);
}

/**
 * Reset the onboarding flow to idle.
 *
 * Clears the `MORPHIC_ONBOARDING_MARKER` sub-key but does NOT touch the user's
 * written prefs (theme/motion/density) at the root level — those reflect
 * autonomous choices and survive a re-onboarding cycle.
 */
export function resetOnboarding(): OnboardingState {
  const newState: OnboardingState = {
    started: false,
    completed: false,
    currentStep: null,
    completedSteps: [],
  };
  currentState = newState;

  const root = readRoot();
  if (MORPHIC_ONBOARDING_MARKER in root) {
    delete root[MORPHIC_ONBOARDING_MARKER];
    writeRoot(root);
  }
  return newState;
}

File: android/morphic/src/main/kotlin/com/theermite/morphic/onboarding/Onboarding.kt
package com.theermite.morphic.onboarding

import com.theermite.morphic.state.MorphicState

/**
 * Sensoriel-AVANT-identité onboarding — pure Kotlin state machine + identity
 * guard (A-5).
 *
 * Mirrors the web `packages/engine/src/onboarding.ts`. Three ordered sensoriel
 * steps (theme → motion → density); identity collection stays forbidden until
 * all three are done. That ordering is the Dignity §a BLOCKING contract: the
 * platform proves it respects the user's comfort BEFORE asking who they are.
 *
 * No Android framework here on purpose — fully measurable by Kover (Critical
 * 95% + MC/DC on the guard). The accessible Compose screen lives next door in
 * [MorphicOnboardingScreen] and is validated in a real app (PET-Android §5.bis).
 */

/** Ordered sensoriel step (closed enum, poka-yoke). `key` mirrors the web step name. */
public enum class OnboardingStep(public val key: String) {
    THEME("theme"),
    MOTION("motion"),
    DENSITY("density"),
    ;

    public companion object {
        /** Canonical order the machine walks through. */
        public val ORDER: List<OnboardingStep> = listOf(THEME, MOTION, DENSITY)
    }
}

/** Immutable snapshot of the onboarding flow. */
public data class OnboardingState(
    /** True once the user has entered the flow. */
    val started: Boolean,
    /** True once all three steps are done (completed or skipped). */
    val completed: Boolean,
    /** Step waiting for input, or null when idle/done. */
    val currentStep: OnboardingStep?,
    /** Steps resolved so far, in order. */
    val completedSteps: List<OnboardingStep>,
)

/**
 * Dignity §a BLOCKING guard — true if AND ONLY IF the user has entered the flow
 * AND finished all three sensoriel steps. Host apps MUST gate any
 * identity-collection UI behind this.
 *
 * Pure function so the MC/DC truth table is testable directly:
 *  - started=T, completed=T → true  (done)
 *  - started=T, completed=F → false (in progress — `completed` flips it)
 *  - started=F, completed=T → false (machine-unreachable — `started` flips it)
 *  - started=F, completed=F → false (idle)
 */
public fun canCollectIdentity(state: OnboardingState): Boolean =
    state.started && state.completed

/**
 * Drives the onboarding flow and writes each resolved choice into [morphic]
 * (parity with the web onboarding, which writes user choices to the prefs).
 *
 * In-memory only: the chosen axis values persist through `MorphicStore` (A-3,
 * host-driven); persisting the "onboarding done" marker is a host concern and
 * out of scope here — same layered split as A-2/A-3.
 */
public class MorphicOnboarding(private val morphic: MorphicState) {
    private var state: OnboardingState = idle()

    /** Current snapshot (in-memory). */
    public fun getState(): OnboardingState = state

    /** Enter the flow. Idempotent: re-entering never wipes progress. */
    public fun start(): OnboardingState {
        if (!state.started) {
            state = OnboardingState(
                started = true,
                completed = false,
                currentStep = OnboardingStep.THEME,
                completedSteps = emptyList(),
            )
        }
        return state
    }

    /**
     * Resolve the current step with a user-chosen value.
     *
     * @throws IllegalStateException if not started, already done, or out of order.
     * @throws IllegalArgumentException if the value is not a valid axis choice
     *   (raised by [MorphicState]); the machine is NOT advanced in that case.
     */
    public fun completeStep(step: OnboardingStep, value: String): OnboardingState =
        advance(step) { apply(step, value) }

    /**
     * Resolve the current step with its documented default
     * (theme=auto, motion=auto, density=comfortable).
     *
     * @throws IllegalStateException if not started, already done, or out of order.
     */
    public fun skipStep(step: OnboardingStep): OnboardingState =
        advance(step) { apply(step, DEFAULTS.getValue(step)) }

    /** Dignity §a guard for the live machine — see the pure [canCollectIdentity]. */
    public fun canCollectIdentity(): Boolean = canCollectIdentity(state)

    /**
     * Reset the ceremony to idle. The user's axis choices in [morphic] are
     * preserved — they reflect autonomous decisions, not the ceremony
     * (RGPD-friendly, parity with web `resetOnboarding`).
     */
    public fun reset(): OnboardingState {
        state = idle()
        return state
    }

    /** Shared guard-then-mutate path. Checks run BEFORE [write] so a failed write leaves state intact. */
    private fun advance(step: OnboardingStep, write: () -> Unit): OnboardingState {
        check(state.started) { "onboarding: not started — call start() first" }
        check(!state.completed) {
            "onboarding: already completed, cannot advance \"${step.key}\""
        }
        check(state.currentStep == step) {
            "onboarding: out of order — expected \"${state.currentStep?.key}\", got \"${step.key}\""
        }
        write()
        val done = state.completedSteps + step
        val allDone = done.size == OnboardingStep.ORDER.size
        state = OnboardingState(
            started = true,
            completed = allDone,
            currentStep = if (allDone) null else OnboardingStep.ORDER[done.size],
            completedSteps = done,
        )
        return state
    }

    private fun apply(step: OnboardingStep, value: String) {
        when (step) {
            OnboardingStep.THEME -> morphic.setTheme(value)
            OnboardingStep.MOTION -> morphic.setMotion(value)
            OnboardingStep.DENSITY -> morphic.setDensity(value)
        }
    }

    private companion object {
        private val DEFAULTS: Map<OnboardingStep, String> = mapOf(
            OnboardingStep.THEME to "auto",
            OnboardingStep.MOTION to "auto",
            OnboardingStep.DENSITY to "comfortable",
        )

        private fun idle(): OnboardingState =
            OnboardingState(started = false, completed = false, currentStep = null, completedSteps = emptyList())
    }
}
